### ISSUE TITLE
json/binary and web browser interoperability issue

### ISSUE THREAD

**Issue by atramos** [aws/chalice#1523]
There is a framework usability issue when you try to set binary and text (json) returns at different endpoints of the same application, e.g. return json in one function and jpeg images in another, for browser consumption (this problem doesn't exist in programmatic clients where you can just set the Accepts: content-type to whatever you want).

The root of the problem is two-fold:
1. `app.api.binary_types` is a global setting in API Gateway.
2. browsers don't send `image/jpeg` as an Accept type,  so you have to use `*/*`

So in order to serve images, you have to set:
`app.api.binary_types =['*/*']`

That's all good and works fine, until you try to invoke a different endpoint returning json.
Then, **boom**:

> (401bd39b-6ba7-481c-b38b-9c54becfaf17) Lambda execution failed with status 200 due to customer function error: Expected bytes type for body with binary Content-Type. Got <class 'str'> type body instead.. Lambda request id: 637d6f64-01bc-444f-9df2-c0259e2a12ab

There is "one weird trick" that lets you return json again:

`return Response(json.dumps(obj).encode(), ...)
`

It took me literally 16 months to get to the bottom of this issue (ok I wasn't working on it the entire time, but still). I had to temporarily split my application into two. Now I had to revisit it because I really need to merge it back into one app.

**TL;DR**

The _expected binary... Got <class 'str'>_ error is so sad, it's depressing. Chalice should automatically convert strings to binary using a default encoding (I heard utf8 is fairly popular) when a binary is expected.

**Comment by jamesls** [comment]
Thanks for reporting, this was also discussed here (https://github.com/aws/chalice/issues/1501), which has more details/background info.

I do think we can improve things by adding the fallback case you suggest, if we need to base64 encode the data, we'll just assume a utf-8 encoding of any str/unicode type as a fallback.  This will let you return data without having to use the `Response(json.dumps(...), ...)` workaround.  I'll go ahead and comment on that issue, and let's track that work over there.
